draft-jmiller-json-web-proof.md

```diff
@@ -70,49 +71,25 @@ when, and only when, they appear in all capitals, as shown here.
 
-# JWP Format
-
-A JWP contains multiple specific payloads, which are always represented within an ordered array. These payloads have the same processing rules applied as they would within JWS.
-
-The individual payloads are often meant to be composed into a single credential, and as such, are most commonly not wholly indpendent from one another, but rather serve as facets of a whole. In keeping with the nomenclature of JWTs, such payloads are said to each represent one or more _claims_.
-
-An individual payload may contain structured information, such as a JSON document representing multiple claims. Another payload might represent a single JWT claim, such as a binary profile image. Other payloads may represent cryptographic values for supporting various proofs, which might also be interpreted as claims.
+Algorithm definitions that support JWPs will be done in separate companion specifications - just as the [JSON Web Algorithms] [@RFC7518] specification does for JWS and JWE [@RFC7516].
 
-## Payload Headers
-
-With multiple payloads, there must also be multiple protected headers in order to safely identify what each payload is and contains.  These headers can themselves easily become a correlation factor if the signer is putting unique values in them or generating unique arragements of headers and their payloads.  In order to minimize these possibilities, JWP places payload headers in the JWK [@!RFC7517] definition so that they are common across all uses of that JWK.
-
-Privacy preserving algorithms have a common pattern of supporting a fixed number of "slots" (also called "messages" or "attributes").  Often they are fixed at the time the public key is created and sometimes they are also pre-defined for a use-case with a specific number of attributes.  This is done to minimize any correlatable signals, to prevent a verifier from categorizing based on if the slots are of variable lengths even when only a subset are revealed in the proof.
+# JWP Format
 
-The key used for a JWP that is identified either by its header or by context MUST have an associated ordered array of payload definitions.  If defined by a JWK, this specification registers a `payloads` parameter with an array value, containing one object for each payload with its associated header values.
+A JWP always contains a protected header along with one or more specific payloads.  The payloads are always serialized and processed as an ordered array.
 
-If the JWK includes a `use` parameter this specification registers a `prf` (proof) value that MUST be used in order for the JWK to be valid for a JWP.
+An individual payload may contain structured information such as a JSON document or be a simple value such as a number, string, or even a binary image.  Some algorithms may support payload values that are cryptographic values such as elliptic curve points or blinded secrets.
```

Comment:
This feels like too many legal values for a payload....

In typescript this type would look like this:

```
type Payload: string | number | Buffer | any; // yikes any? really?
```

This will cause security issues later.

IMO, payload should be JSON... if you can't serialize or deserialize JSON (IoT / CBOR/ pure binary use cases)....you probably won't like base64url encoding either... 

Alternatively, we might consider CBOR instead of JSON, but I really don't like the payload not having a single know type.

Reply:
JWS supports an unlimited amount of legal values for a payload - in that it's simply signing an array of bytes.  Similarly, the JWP layer should know nothing about the content of ifs payloads, other than their representations as arrays of bytes.

Reply:
I am in favor of:

```
type PayloadOptions = Buffer;
type Payload = base64url (Buffer). base64url (PayloadOptions).
type Payloads = Array<Payload>;
```

implementers shouldn't have to care about the payload content, the alg registry should tell you how to care about options or payloads.

Reply:
This is an unfortunate complexity I think we'll have to work through and deal with in some form.

The JWP _algorithm interface_ simply *must* support rich types.  JWP itself as a container format absolutely treats each payload as an array of bytes, but during the act of signing, proof generation, and proof validation the algorithm has to have the ability to accept rich types.

These rich interfaces are language/library level and not part of the protocol, but they are important to support required capabilities such as blinded values (algo has to be given an elliptic curve point), predicate proofs (algo requires a number, set member value, etc).

There's no reason to make these encodings be part of the protocol, they really are an implementation language level detail and algorithm specific.  There's not even an interop issue that I can tell since the algorithms that support these have very specific ways they use them via a hash2curve or other such cryptographic transformation.

I definitely need help explaining this and adding better text around it, so I welcome the discussions :)

Reply:
Let's continue this in issue #8